### ISSUE TITLE
Implementing parameter sharing (Universal Transformers)

### ISSUE THREAD

**Issue by sarthmit** [pytorch/fairseq#3261]
#### What is your question?
I am trying to set up Universal Transformer in the codebase and wanted some help and insight on it. Basically Universal Transformers are just normal transformer models with parameter sharing across layers. 

#### Code
I am changing the [following line](https://github.com/pytorch/fairseq/blob/4cf7d76114d50008cdd98a7fde250d4ef99b66fe/fairseq/models/transformer.py#L374) to have the same initialized layer within the loop in `self.layers.extend` for both the encoder and the decoder. Is that all the change that is required to get it set up because I am not able to get good results with it.

**Comment by lematt1991** [comment]
Can you provide a branch to the code with your changes?

**Comment by sarthmit** [comment]
You can find the model defined here

[https://gist.github.com/sarthmit/e2c92536138dc227c3e3a48e5ddce8a7](https://gist.github.com/sarthmit/e2c92536138dc227c3e3a48e5ddce8a7)

This is essentially the only change I make to the model definition. The main changes can be found on lines #377-380 and #631-637 where I basically just repeat the same layer ```num_layers``` number of times.

**Comment by lematt1991** [comment]
Yes, this should be all you need to do to share weights across layers.

**Comment by sarthmit** [comment]
Thank you! Is there anything that needs to be changed in the optimization procedure?

I had a few follow-up questions. 
1. Is the fairseq repository going to set up a Universal Transformers codebase or are there any plans for it?
2. How are the hyper-parameters set up as opposed to the the setup in Google's tensor2tensor codebase? In particular, how is batch size and learning rate parameterized?

Thank you very much for the prompt response!

**Comment by lematt1991** [comment]
> Is there anything that needs to be changed in the optimization procedure?  

There shouldn't be based on your description of UT, although I'm not very familiar with the paper, so I don't know if they do anything special here.

> Is the fairseq repository going to set up a Universal Transformers codebase or are there any plans for it?

Not that I know of.

> How are the hyper-parameters set up as opposed to the the setup in Google's tensor2tensor codebase? In particular, how is batch size and learning rate parameterized?

You can find details of hyper-parameters [here](https://github.com/pytorch/fairseq/blob/master/fairseq/dataclass/configs.py).  batch size is parameterized in two ways: `max_tokens` or `batch_size`.  `max_tokens` says that a batch can contain at most `N` tokens, but the number of examples (sentences) is allowed to vary.  `batch_size` uses a fixed number of examples (sentences) in each batch.

**Comment by sarthmit** [comment]
Thanks, how about the learning rate? Because in tensor2tensor I saw considerably large learning rates - [https://github.com/tensorflow/tensor2tensor/blob/9902e885fff48d610034b52153f7769081fcf2e0/tensor2tensor/models/transformer.py#L1969](https://github.com/tensorflow/tensor2tensor/blob/9902e885fff48d610034b52153f7769081fcf2e0/tensor2tensor/models/transformer.py#L1969) whereas in fairseq I find learning rates around 7.5e-4

**Comment by lematt1991** [comment]
You can find details about `lr` in the same file I linked.  The default is 0.25, but you can change this via the command line (or a config file).  Many other hyper-parameters will likely differ, so you'll want to track them all down and get consistent inputs when comparing the two implementations.

**Comment by sarthmit** [comment]
Could you point me to where it interacts with the lr schedulers (eg. with inverse sqrt)?

**Comment by lematt1991** [comment]
You can find the various lr schedulers [here](https://github.com/pytorch/fairseq/tree/master/fairseq/optim/lr_scheduler)

**Comment by sarthmit** [comment]
Thank you very much!

**Comment by takase** [comment]
Hi, I notice this issue thread now.
Did @sarthmit solve the problem of your Universal Transformers?
I think the generated results are not good due to [buffers in multi head attentions in the decoder part](https://github.com/pytorch/fairseq/blob/ab560669cd9baaa4009e1fd01c970f8ffccd1ee0/fairseq/modules/transformer_layer.py#L320), aren't they?
I implemented Universal Transformers in the same way before, and had a trouble in the above part.

**Comment by sarthmit** [comment]
Hello @takase 
I am not able to get results comparative to the Universal Transformers paper but I am not sure why.
Could you elaborate on the buffers part? In particular, what are they, how are they implemented and why do they affect the Universal Transformer in a bad way?

@lematt1991 your opinion on this would be super helpful too!

**Comment by takase** [comment]
Could you tell me the parameter size and the valid loss value of newstest 2013?
I achieved 4.233 valid loss with 24M parameters, and it achieve 24.71 BLEU score in newstest 2014 with SacreBLEU.

If you can train a model whose performance is competitive with the above valid loss, and the generated results are not good, the buffer parts harm generations.

In the generation step, the multi head attention reuses the computational result at previous time step as the buffer (in my understanding).
Your implementation copies layers, and thus, the decoders use the same computational results (the same buffers) in the generation step.

The simplest way; train a model with your implemented universal transformer, and then use the basic transformer in the generation step.
Otherwise, modify the code to copy only parameters instead of layers.

**Comment by sarthmit** [comment]
I am getting 2.6 valid loss with around 21.5 BLEU score on English to German newstest2013 data. Training is done on WMT14 data and the model has around 50M params.

Would you be willing to share your code with me? The model is just a copy of basic Transformer with the same layer repeated so how would using a normal transformer solve the problem?

**Comment by takase** [comment]
I modified various parts of my code, so cannot share my code directly.
I add the function to copy parameters such as:

`def CopyParams(base_layer, target_layer, decoder=False):
    target_layer.self_attn.k_proj = base_layer.self_attn.k_proj
    target_layer.self_attn.v_proj = base_layer.self_attn.v_proj
    target_layer.self_attn.q_proj = base_layer.self_attn.q_proj
    target_layer.self_attn.out_proj = base_layer.self_attn.out_proj
    if base_layer.self_attn.bias_k is not None:
        target_layer.self_attn.bias_k = base_layer.self_attn.bias_k
    if base_layer.self_attn.bias_v is not None:
        target_layer.self_attn.bias_v = base_layer.self_attn.bias_v
    target_layer.fc1 = base_layer.fc1
    target_layer.fc2 = base_layer.fc2
    target_layer.self_attn_layer_norm = base_layer.self_attn_layer_norm
    target_layer.final_layer_norm = base_layer.final_layer_norm
    if decoder:
        target_layer.encoder_attn.k_proj = base_layer.encoder_attn.k_proj
        target_layer.encoder_attn.v_proj = base_layer.encoder_attn.v_proj
        target_layer.encoder_attn.q_proj = base_layer.encoder_attn.q_proj
        target_layer.encoder_attn.out_proj = base_layer.encoder_attn.out_proj
        if base_layer.encoder_attn.bias_k is not None:
            target_layer.encoder_attn.bias_k = base_layer.encoder_attn.bias_k
        if base_layer.encoder_attn.bias_v is not None:
            target_layer.encoder_attn.bias_v = base_layer.encoder_attn.bias_v
        target_layer.encoder_attn_layer_norm = base_layer.encoder_attn_layer_norm
    return target_layer`

In addition, copying parameters in layer construction as following:

`            for _ in range(args.encoder_layers):
                 if _ == 0:
                    base_encoder_layer = self.build_encoder_layer(args)
                    self.layers.extend([base_encoder_layer])
                else:
                    encoder_layer = self.build_encoder_layer(args)
                    encoder_layer = CopyParams(base_encoder_layer, encoder_layer)
                    self.layers.extend([encoder_layer])`

**Comment by takase** [comment]
I'm sorry the indent of the above code is broken.

**Comment by sarthmit** [comment]
What is the difference that you saw in the evaluation with and without copying weights?

Does it also affect training? For some reason I have a good validation loss but the bleu score is poor, do you have any idea why?

**Comment by takase** [comment]
As described, since the multi head attention reuses the computational result at previous time step as the buffer (in my understanding), BLEU score is poor even if a model achieve a good validation loss.
https://github.com/pytorch/fairseq/issues/3261#issuecomment-784822751

If we copy parameters instead of the layer class, the code doesn't use the buffers in unexpected time steps.
The buffers don't harm training, and thus, it was difficult for me to fix it.

**Comment by sarthmit** [comment]
I shouldn't be doing the copying of parameters in the model definition though, right? Because that would just initialize the different layers to be the same, but the training would destroy it.

**Comment by takase** [comment]
You shouldn't copy layers including all functions.
Instead, you should copy parameters instead of whole layers, in my opinion.

**Comment by sarthmit** [comment]
Ohh, I see what you mean! Does that solve the issue? Are you able to replicate the paper's results with that?
That would be cool (I have spent so much time on this without success lol)

**Comment by takase** [comment]
Thanks!
Yes, my implemented Universal Transformer outperformed the basic Transformer result if I used the same parameter size for them.

**Comment by sarthmit** [comment]
Thanks @takase!
That solved the problem!
